js/style/style_layer.js

```diff
@@ -34,44 +34,56 @@ function StyleLayer(layer, refLayer) {
     this.minzoom = (refLayer || layer).minzoom;
     this.maxzoom = (refLayer || layer).maxzoom;
     this.filter = (refLayer || layer).filter;
-    this.layout = (refLayer || layer).layout;
-
-    this._paintDeclarations = {}; // {[class name]: { [property name]: StyleDeclaration }}
-    this._paintTransitions = {}; // {[class name]: { [property name]: StyleTransitionOptions }}
-    this._paintTransitions = {}; // { [property name]: StyleDeclaration }
+    this.interactive = (refLayer || layer).interactive;
 
     this._paintSpecifications = StyleSpecification['paint_' + this.type];
     this._layoutSpecifications = StyleSpecification['layout_' + this.type];
+
+    // Resolve paint declarations
+    this._paintDeclarations = {}; // { [class name]: { [property name]: StyleDeclaration }}
+    this._paintTransitions = {}; // { [class name]: { [property name]: StyleTransitionOptions }}
+    for (var key in this._layer) {
+        var match = key.match(/^paint(?:\.(.*))?$/);
+        if (match) {
+            var klass = match[1] || '';
+            for (var name in this._layer[key]) {
+                this.setPaintProperty(name, this._layer[key][name], klass);
+            }
+        }
+    }
+
+    // Resolve layout declarations
+    this._layoutDeclarations = {}; // { [property name]: StyleDeclaration }
+    if (!this._refLayer) {
+        for (name in this._layer.layout) {
+            this.setLayoutProperty(name, this._layer.layout[name]);
+        }
+    } else {
+        this._layoutDeclarations = this._refLayer._layoutDeclarations;
+    }
 }
 
 StyleLayer.prototype = {
-    resolveLayout: function() {
-        if (!this.ref) {
-            this.layout = new LayoutProperties[this.type](this._layer.layout);
-        }
-    },
 
     setLayoutProperty: function(name, value) {
-        if (value == null) {
```

Comment:
Weirdly, fixing this behavior caused some [unit tests to break](https://github.com/mapbox/mapbox-gl-js/pull/1977/files#diff-df9dac9f0f6edf6f9675a223d7687c17R88). I think think we should consider the tests wrong and the intended behavior (i.e. your proposed change) correct.